### PR TITLE
Future proof the configuration

### PR DESCRIPTION
**Pull request by TristanCacqueray**
Fixes https://github.com/change-metrics/monocle/issues/466

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
Todo:

- [ ] Adapt the python config loader to transparently migrate the old configuration.

**Comment by TristanCacqueray** [comment]
I have to say, it's so convenient to have the configuration schema in the host language date type, it makes such refactoring very safe :-)

**Comment by TristanCacqueray** [comment]
We might want to carefully test the change, but I think this should be good to go now.

File: README.md [hunk 1]
```diff
@@ -343,10 +345,11 @@ Check the OpenAPI definitions for tasks data endpoints: [Monocle OpenAPI][monocl
 ```YAML
 tenants:
   - index: default
-    task_crawlers:
+    crawlers_api_key: 1a2b3c4d5e
+    crawlers:
       - name: crawler_name
         updated_since: "2020-01-01"
-        api_key: 1a2b3c4d5e
+        provider: TaskDataProvider
```
Comment (morucci) [inline README.md:352]: Could we have a complete example for BugZilla ?

File: README.md [hunk 2]
```diff
@@ -362,59 +365,67 @@ the initial date.
...
-      loop_delay: 600
-      gerrit_repositories:
-        - name: "^openstack/.*"
-          updated_since: "2021-01-01"
-          base_url: https://review.opendev.org/
-        - name: "^openstack/.*"
-          updated_since: "2021-01-01"
-          base_url: https://review.rdoproject.org/r/
-          prefix: "rdo/"
+        provider: TaskDataProvider
```
Comment (morucci) [inline README.md:401]: Same here

File: README.md [hunk 3]
```diff
@@ -90,14 +90,12 @@ Then create the config file `etc/config.yaml`. Here is an example your could sta
...
-      github_orgs:
-        - name: tektoncd
-          repository: pipeline
-          updated_since: "2020-05-01"
-          token: <github_token>
-          base_url: https://github.com
+    crawlers:
+      - name: github-tektoncd
+        provider:
+          github_token: <github_token>
```
Comment (morucci) [inline README.md:96]: github_url is optional now, but it seems the legacy code do not provide a fallback to github.com

File: monocle/config.py [hunk 4]
```diff
@@ -25,71 +27,71 @@
...
+            "type": "object",
+            "required": ["name", "update_since", "provider"],
+            "properties": {
+                "name": {"type": "string"},
+                "update_since": {"type": "string", "format": "date"},
+                "provider": {
+                    "oneOf": [
+                        {"$ref": "#/definitions/gerrit_provider"},
+                        {"$ref": "#/definitions/github_provider"},
+                        {"$ref": "#/definitions/gitlab_provider"},
```
Comment (morucci) [inline monocle/config.py:89]: Do we need the GitLab crawler here ?

File: monocle/config.py [hunk 5]
```diff
@@ -170,66 +150,154 @@
...
+            for repo in provider.get("github_repositories", [None]):
+                tenant["crawler"]["github_orgs"].append(
+                    dict(
+                        updated_since=crawler["update_since"],
+                        name=provider["github_organization"],
+                        token=provider["github_token"],
+                        repository=repo,
+                        base_url=provider["github_url"]
+                        if provider.get("github_url")
+                        else "https://github.com",
```
Comment (morucci) [inline monocle/config.py:207]: ok

File: monocle/config.py [hunk 6]
```diff
@@ -170,66 +150,154 @@
...
+
+
+def upgrade(tenant):
+    """Take a v0.9 configuration and return a v1.0"""
+    crawlers = tenant.get("crawlers", [])
+
+    # Add github crawlers
+    legacy_crawler = tenant.pop("crawler", {})
+    for crawler in legacy_crawler.get("github_orgs", []):
+        if crawler.get("base_url", "") == "https://github.com":
```
Comment (morucci) [inline monocle/config.py:235]: ok

File: README.md (hunk 3 above)
Comment (morucci) [reply README.md:96]: Ok saw it later in the migration tool

**Review by morucci — APPROVED** [review]

File: monocle/config.py (hunk 4 above)
Comment (TristanCacqueray) [reply monocle/config.py:89]: Turns out json schema is pretty flexible, and we might omit this list.

File: README.md (hunk 1 above)
Comment (TristanCacqueray) [reply README.md:352]: We could, but this is actually not implemented, we'll need to migrate the lentille-bugzilla first.

**Comment by TristanCacqueray** [comment]
I think we could merge that now.

**Pull request merged**